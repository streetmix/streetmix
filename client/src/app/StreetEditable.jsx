import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { DropTarget } from 'react-dnd'
import { CSSTransition, TransitionGroup } from 'react-transition-group'
import { flow } from '../util/flow'
import Segment from '../segments/Segment'
import {
  TILE_SIZE,
  DRAGGING_MOVE_HOLE_WIDTH,
  DRAGGING_TYPE_RESIZE
} from '../segments/constants'
import { cancelSegmentResizeTransitions } from '../segments/resizing'
import {
  Types,
  canvasTarget,
  collectDropTarget,
  makeSpaceBetweenSegments,
  isSegmentWithinCanvas
} from '../segments/drag_and_drop'

export class StreetEditable extends React.Component {
  static propTypes = {
    // Provided by parent
    resizeType: PropTypes.number,
    setBuildingWidth: PropTypes.func.isRequired,
    updatePerspective: PropTypes.func.isRequired,
    draggingType: PropTypes.number,

    // Provided by store
    street: PropTypes.object.isRequired,
    draggingState: PropTypes.object,

    // Provided by DropTarget
    connectDropTarget: PropTypes.func
  }

  // Internal "state", but does not affect renders, so it is not React state
  withinCanvas = null

  // Placeholder for a ref.
  // TODO: Upgrade to createRef(), but this is currently broken when placed on
  // an element inside of react-dnd's `connectDragSource`.
  // Info: https://github.com/react-dnd/react-dnd/issues/998
  streetSectionEditable = null

  componentDidMount () {
    this.props.setBuildingWidth(this.streetSectionEditable)
  }

  componentDidUpdate (prevProps) {
    const { resizeType, draggingState } = this.props

    if (
      (resizeType && !prevProps.resizeType) ||
      (prevProps.draggingType === DRAGGING_TYPE_RESIZE &&
        !this.props.draggingType)
    ) {
      this.props.setBuildingWidth(this.streetSectionEditable)
    }

    if (
      prevProps.street.id !== this.props.street.id ||
      prevProps.street.width !== this.props.street.width
    ) {
      cancelSegmentResizeTransitions()
    }

    const dragEvents = ['dragover', 'touchmove']
    if (!prevProps.draggingState && draggingState) {
      dragEvents.forEach((type) => {
        window.addEventListener(type, this.updateWithinCanvas)
      })
    } else if (prevProps.draggingState && !draggingState) {
      dragEvents.forEach((type) => {
        window.removeEventListener(type, this.updateWithinCanvas)
      })
    }
  }

  updateWithinCanvas = (event) => {
    const withinCanvas = isSegmentWithinCanvas(
      event,
      this.streetSectionEditable
    )

    if (withinCanvas) {
      document.body.classList.remove('not-within-canvas')
    } else {
      document.body.classList.add('not-within-canvas')
    }

    if (this.withinCanvas !== withinCanvas) {
      this.withinCanvas = withinCanvas
    }
  }

  updateSegmentData = (ref, dataNo, segmentPos) => {
    const { segments } = this.props.street
    const segment = segments[dataNo]

    if (segment) {
      ref.dataNo = dataNo
      ref.savedLeft = Math.round(segmentPos)
    }
  }

  handleSwitchSegmentAway = (el) => {
    el.classList.add('create')
    el.style.left = el.savedLeft + 'px'

    this.props.updatePerspective(el)
  }

  calculateSegmentPos = (dataNo) => {
    const { segments, remainingWidth } = this.props.street
    const { draggingState } = this.props

    let currPos = 0

    for (let i = 0; i < dataNo; i++) {
      const width =
        draggingState && draggingState.draggedSegment === i
          ? 0
          : segments[i].width * TILE_SIZE
      currPos += width
    }

    let mainLeft = remainingWidth
    if (draggingState && segments[draggingState.draggedSegment] !== undefined) {
      const draggedWidth = segments[draggingState.draggedSegment].width || 0
      mainLeft += draggedWidth
    }

    mainLeft = (mainLeft * TILE_SIZE) / 2

    if (draggingState && this.withinCanvas) {
      mainLeft -= DRAGGING_MOVE_HOLE_WIDTH
      const spaceBetweenSegments = makeSpaceBetweenSegments(
        dataNo,
        draggingState
      )
      return mainLeft + currPos + spaceBetweenSegments
    } else {
      return mainLeft + currPos
    }
  }

  onExitAnimations = (child) => {
    return React.cloneElement(child, {
      exit: !this.props.street.immediateRemoval
    })
  }

  renderStreetSegments = () => {
    const { segments, units, immediateRemoval } = this.props.street
    const streetId = this.props.street.id

    return segments.map((segment, i) => {
      const segmentPos = this.calculateSegmentPos(i)

      const segmentEl = (
        <CSSTransition
          key={`${streetId}.${segment.id}`}
          timeout={250}
          classNames="switching-away"
          exit={!immediateRemoval}
          onExit={this.handleSwitchSegmentAway}
          unmountOnExit={true}
        >
          <Segment
            dataNo={i}
            segment={{ ...segment }}
            actualWidth={segment.width}
            units={units}
            segmentPos={segmentPos}
            updateSegmentData={this.updateSegmentData}
            updatePerspective={this.props.updatePerspective}
          />
        </CSSTransition>
      )

      return segmentEl
    })
  }

  render () {
    const { connectDropTarget } = this.props
    const style = {
      width: this.props.street.width * TILE_SIZE + 'px'
    }

    return connectDropTarget(
      <div
        id="street-section-editable"
        key={this.props.street.id}
        style={style}
        ref={(ref) => {
          this.streetSectionEditable = ref
        }}
      >
        <TransitionGroup
          key={this.props.street.id}
          component={null}
          enter={false}
          childFactory={this.onExitAnimations}
        >
          {this.renderStreetSegments()}
        </TransitionGroup>
      </div>
    )
  }
}

function mapStateToProps (state) {
  return {
    street: state.street,
    draggingState: state.ui.draggingState
  }
}

export default flow([
  DropTarget(
    [Types.SEGMENT, Types.PALETTE_SEGMENT],
    canvasTarget,
    collectDropTarget
  ),
  connect(mapStateToProps)
])(StreetEditable)
