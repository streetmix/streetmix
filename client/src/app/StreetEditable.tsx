import React, { useEffect, useRef, createRef, cloneElement } from 'react'
import { useDrop } from 'react-dnd'
import { CSSTransition, TransitionGroup } from 'react-transition-group'

import { useSelector } from '~/src/store/hooks'
import { usePrevious } from '~/src/util/usePrevious'
import type { DraggingState } from '~/src/types'
import Segment from '../segments/Segment'
import {
  TILE_SIZE,
  DRAGGING_MOVE_HOLE_WIDTH,
  DRAGGING_TYPE_RESIZE
} from '../segments/constants'
import { cancelSegmentResizeTransitions } from '../segments/resizing'
import {
  isSegmentWithinCanvas,
  createStreetDropTargetSpec
} from '../segments/drag_and_drop'

/**
 * Calculates the gap shown before or after slices while dragging another slice
 */
function makeSpaceBetweenSlices (
  sliceIndex: number,
  draggingState: DraggingState
): number {
  const { segmentBeforeEl, segmentAfterEl } = draggingState

  let gap = 0

  if (segmentBeforeEl !== undefined && sliceIndex >= segmentBeforeEl) {
    gap += DRAGGING_MOVE_HOLE_WIDTH

    if (segmentAfterEl === undefined) {
      gap += DRAGGING_MOVE_HOLE_WIDTH
    }
  }

  if (segmentAfterEl !== undefined && sliceIndex > segmentAfterEl) {
    gap += DRAGGING_MOVE_HOLE_WIDTH

    if (segmentBeforeEl === undefined) {
      gap += DRAGGING_MOVE_HOLE_WIDTH
    }
  }

  return gap
}

interface StreetEditableProps {
  resizeType: number | null
  setBuildingWidth: (node: HTMLDivElement | null) => void
  updatePerspective: (el: HTMLElement | null) => void
  draggingType?: number
}

function StreetEditable (props: StreetEditableProps): React.ReactElement {
  const { resizeType, setBuildingWidth, updatePerspective, draggingType } =
    props
  const street = useSelector((state) => state.street)
  const draggingState = useSelector((state) => state.ui.draggingState)

  // Internal "state", but does not affect renders, so it is not React state
  const withinCanvas = useRef<boolean>(false)
  const streetSectionEditable = useRef<HTMLDivElement>(null)

  // According to "rule of hooks", useRef() must not be called in a loop
  // This is a top-level container to manage a list of refs as a workaround
  // for CSSTransition's reliance on deprecated findDOMNode
  const childRefs = useRef<Record<string, React.RefObject<HTMLDivElement>>>({})

  // Keep previous state for comparisons (ported from legacy behavior)
  const prevProps = usePrevious({
    resizeType,
    draggingType,
    street,
    draggingState
  })

  // Set up drop target
  const dropTargetSpec = createStreetDropTargetSpec(
    street,
    streetSectionEditable
  )
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [collectedProps, drop] = useDrop(dropTargetSpec)

  useEffect(() => {
    if (!prevProps?.draggingState && draggingState) {
      window.addEventListener('dragover', updateWithinCanvas)
      window.addEventListener('touchmove', updateWithinCanvas)
    } else if (prevProps?.draggingState && !draggingState) {
      window.removeEventListener('dragover', updateWithinCanvas)
      window.removeEventListener('touchmove', updateWithinCanvas)
    }

    // Cleanup
    return () => {
      window.removeEventListener('dragover', updateWithinCanvas)
      window.removeEventListener('touchmove', updateWithinCanvas)
    }
  }, [draggingState])

  useEffect(() => {
    if (prevProps === null || prevProps === undefined) return
    if (
      (resizeType !== undefined && prevProps.resizeType !== undefined) ??
      (prevProps.draggingType === DRAGGING_TYPE_RESIZE &&
        draggingType !== undefined)
    ) {
      setBuildingWidth(streetSectionEditable.current)
    }
  }, [resizeType, draggingType])

  useEffect(() => {
    if (
      prevProps?.street.id !== street.id ||
      prevProps?.street.width !== street.width
    ) {
      cancelSegmentResizeTransitions()
    }
  }, [street.id, street.width])

  function updateWithinCanvas (event: MouseEvent | TouchEvent): void {
    const newValue = isSegmentWithinCanvas(event, streetSectionEditable.current)

    if (newValue) {
      document.body.classList.remove('not-within-canvas')
    } else {
      document.body.classList.add('not-within-canvas')
    }

    if (withinCanvas.current !== newValue) {
      withinCanvas.current = newValue
    }
  }

  function handleSwitchSliceAway (el: HTMLDivElement, sliceIndex: number): void {
    // Targeting first child node instead of el because of wrapper div workaround
    // for CSSTransition
    const childNode = el.firstChild as HTMLDivElement
    if (childNode === null) return

    const left = calculateSlicePosition(sliceIndex)
    childNode.style.left = `${left}px`

    updatePerspective(childNode)
  }

  function calculateSlicePosition (sliceIndex: number): number {
    const { segments, remainingWidth } = street

    let currPos = 0

    for (let i = 0; i < sliceIndex; i++) {
      const width =
        draggingState && draggingState.draggedSegment === i
          ? 0
          : segments[i].width * TILE_SIZE
      currPos += width
    }

    let mainLeft = remainingWidth
    if (draggingState && segments[draggingState.draggedSegment] !== undefined) {
      const draggedWidth = segments[draggingState.draggedSegment].width || 0
      mainLeft += draggedWidth
    }

    mainLeft = (mainLeft * TILE_SIZE) / 2

    if (draggingState && withinCanvas.current) {
      mainLeft -= DRAGGING_MOVE_HOLE_WIDTH
      const gap = makeSpaceBetweenSlices(sliceIndex, draggingState)
      return mainLeft + currPos + gap
    } else {
      return mainLeft + currPos
    }
  }

  function onExitAnimations (child: React.ReactElement): React.ReactElement {
    return cloneElement(child, {
      exit: !street.immediateRemoval
    })
  }

  function renderStreetSegments (): React.ReactNode {
    const { segments, units, immediateRemoval } = street
    const streetId = street.id

    return segments.map((segment, i) => {
      const segmentLeft = calculateSlicePosition(i)
      const key = `${streetId}.${segment.id}`
      // Refs are created with createRef and then stored in parent `refs`
      const ref = createRef<HTMLDivElement>()
      childRefs.current[key] = ref

      const segmentEl = (
        <CSSTransition
          key={key}
          timeout={250}
          classNames="slice-remove"
          exit={!immediateRemoval}
          onExit={() => {
            if (ref.current === null) return
            handleSwitchSliceAway(ref.current, i)
          }}
          unmountOnExit
          nodeRef={ref}
        >
          {/* This wrapper element is a workaround for CSSTransition depending on
              findDOMNode, which is deprecated. We just need a DOM element to
              attach a ref to. This introduces a lot of bugs, unfortunately. */}
          <div ref={ref}>
            <Segment
              sliceIndex={i}
              segment={{ ...segment }}
              units={units}
              segmentLeft={segmentLeft}
            />
          </div>
        </CSSTransition>
      )

      return segmentEl
    })
  }

  const style = {
    width: street.width * TILE_SIZE + 'px'
  }

  return (
    <div
      id="street-section-editable"
      key={street.id}
      style={style}
      ref={streetSectionEditable}
    >
      <div style={{ width: '100%', height: '100%' }} ref={drop}>
        <TransitionGroup
          key={street.id}
          component={null}
          enter={false}
          childFactory={onExitAnimations}
        >
          {renderStreetSegments()}
        </TransitionGroup>
      </div>
    </div>
  )
}

export default StreetEditable
