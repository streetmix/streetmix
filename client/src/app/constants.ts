export const JUST_SIGNED_IN_PATH = '/services/auth/just-signed-in'

// Path segments
export const URL_NEW_STREET = '/new'
export const URL_NEW_STREET_COPY_LAST = '/copy-last'
export const URL_GLOBAL_GALLERY = '/gallery'
export const URL_ERROR = '/error'
export const URL_HELP = '/help'
export const URL_SURVEY_FINISHED = '/survey-finished'

// Error fragments that occur after /error/
export const URL_ERROR_NO_TWITTER_REQUEST_TOKEN = 'no-twitter-request-token'
export const URL_ERROR_NO_TWITTER_ACCESS_TOKEN = 'no-twitter-access-token'
export const URL_ERROR_NO_ACCESS_TOKEN = 'no-access-token'
export const URL_ERROR_AUTHENTICATION_API_PROBLEM = 'authentication-api-problem'
export const URL_ERROR_ACCESS_DENIED = 'access-denied'

export const URL_EXAMPLE_STREET = '/streetmix/7'

// Since URLs like “streetmix.net/new” are reserved, but we still want
// @new to be able to use Streetmix, we prefix any reserved URLs with ~
export const RESERVED_URLS = [
  '/services',
  URL_NEW_STREET,
  URL_NEW_STREET_COPY_LAST,
  URL_GLOBAL_GALLERY,
  URL_ERROR,
  URL_HELP,
  URL_SURVEY_FINISHED,
  '/streets',
  '/terms-of-service',
  '/privacy-policy',
  '/map',
  '/survey'
]

export const URL_RESERVED_PREFIX = '~'

// Color modes
export const COLOR_MODE_LIGHT = 'light'
export const COLOR_MODE_DARK = 'dark'
export const COLOR_MODE_AUTO = 'auto'

export type ColorModes = 'light' | 'dark' | 'auto'
