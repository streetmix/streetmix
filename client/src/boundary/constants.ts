export const BOUNDARY_SPACE = 360
export const BOUNDARY_LEFT_POSITION = 'left'
export const BOUNDARY_RIGHT_POSITION = 'right'

export const MAX_BUILDING_FLOORS = 20
