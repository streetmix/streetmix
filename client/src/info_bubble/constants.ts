export const INFO_BUBBLE_TYPE_SEGMENT = 1
export const INFO_BUBBLE_TYPE_LEFT_BUILDING = 2
export const INFO_BUBBLE_TYPE_RIGHT_BUILDING = 3
