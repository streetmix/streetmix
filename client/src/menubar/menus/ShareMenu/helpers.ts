import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
