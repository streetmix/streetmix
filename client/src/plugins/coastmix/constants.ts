export const SEA_LEVEL_YEAR_2030 = 2030
export const SEA_LEVEL_YEAR_2050 = 2050
export const SEA_LEVEL_YEAR_2070 = 2070
