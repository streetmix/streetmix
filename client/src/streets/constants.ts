export const NEW_STREET_DEFAULT = 1
export const NEW_STREET_EMPTY = 2

export const MIN_CUSTOM_STREET_WIDTH = 3 // meters
export const MAX_CUSTOM_STREET_WIDTH = 120 // meters
export const MIN_CUSTOM_STREET_WIDTH_IMPERIAL = 10 * 0.3048 // feet
export const MAX_CUSTOM_STREET_WIDTH_IMPERIAL = 400 * 0.3048 // feet

export const DEFAULT_CAPACITY_SOURCE = 'giz'
