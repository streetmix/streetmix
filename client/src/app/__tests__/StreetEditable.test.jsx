import React from 'react'
import { vi } from 'vitest'
import { userEvent } from '@testing-library/user-event'

import { render, waitFor } from '~/test/helpers/render'
import { SETTINGS_UNITS_METRIC } from '../../users/constants'
import StreetEditable from '../StreetEditable'

describe('StreetEditable', () => {
  beforeEach(() => {
    vi.resetModules()
  })

  const setBuildingWidth = vi.fn()
  const updatePerspective = () => {}
  const type = 'streetcar'
  const variantString = 'inbound|regular'
  const segment = { variantString, id: '1', width: 400, type }

  it('calls setBuildingsWidth', () => {
    const street = {
      segments: [segment],
      width: 100,
      units: SETTINGS_UNITS_METRIC
    }

    render(
      <StreetEditable
        setBuildingWidth={setBuildingWidth}
        updatePerspective={updatePerspective}
      />,
      { street }
    )

    expect(setBuildingWidth).toHaveBeenCalledTimes(1)
  })

  describe('segment warnings', () => {
    describe('too large', () => {
      it('Pressing `+` does not increase the width of the segment', async () => {
        const street = {
          segments: [segment],
          width: 120,
          units: SETTINGS_UNITS_METRIC
        }

        const { getByTestId, store, container, asFragment } = render(
          <StreetEditable
            setBuildingWidth={setBuildingWidth}
            updatePerspective={updatePerspective}
          />,
          { initialState: { street } }
        )

        await userEvent.hover(getByTestId('segment'))
        await userEvent.type(container, '+')
        await waitFor(
          () => {
            expect(store.getState().street.segments[0].width).toEqual(120)
            expect(store.getState().street.segments[0].warnings).toEqual([
              undefined,
              false,
              false,
              true,
              false
            ])
            expect(asFragment()).toMatchSnapshot()
          },
          { container }
        )
      })
    })
  })
})
