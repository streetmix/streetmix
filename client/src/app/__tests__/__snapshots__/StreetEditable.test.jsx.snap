// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StreetEditable > segment warnings > too large > Pressing \`+\` does not increase the width of the segment 1`] = `
<DocumentFragment>
  <div
    id="street-section-editable"
    style="width: 4724.4094488188975px;"
  >
    <div
      class="segment hover show-drag-handles warning"
      data-testid="segment"
      style="width: 4724.4094488188975px; z-index: 15; transform: translateX(0px);"
    >
      <div
        class="segment-label-container"
      >
        <div
          class="segment-grid units-metric"
        />
        <div
          class="segment-width"
        >
          <span>
            120 m
          </span>
        </div>
        <div
          class="segment-label"
        >
          <p
            class="segment-name"
          >
            Streetcar
          </p>
          <p
            class="segment-capacity"
          >
            18,000 people/hr
          </p>
        </div>
      </div>
      <span
        class="drag-handle drag-handle-left"
      >
        <svg
          class="svg-inline--fa fa-chevron-left"
        />
      </span>
      <span
        class="drag-handle drag-handle-right"
      >
        <svg
          class="svg-inline--fa fa-chevron-right"
        />
      </span>
      <div
        class="segment-canvas-container"
      >
        <canvas
          class="segment-image"
          height="445"
          style="width: 4724px; height: 445px; left: 0px;"
          width="4724"
        />
      </div>
      <div
        class="hover-bk"
      />
    </div>
  </div>
</DocumentFragment>
`;
